### PR TITLE
Add singularity config fakeroot command to manage fakeroot user mappings

### LINKED ISSUE #4345: Add singularity config fakeroot command
This is part of #3944. In order to ease management of fakeroot allowed users, we will introduce a command `singularity config fakeroot` in order to add/remove/disable users with :

```
singularity config fakeroot --add user
singularity config fakeroot --remove user
singularity config fakeroot --disable user
singularity config fakeroot --enable user
```

Those commands will automatically take care of user mappings allocation to minimize mistakes (UID/GID overlapping) which could be introduced by doing that manually

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

This PR adds a new `singularity config fakeroot` command to easy fakeroot user mappings management.

**This fixes or addresses the following GitHub issues:**

- Fixes #4345 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: cmd/internal/cli/actions_linux.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func EnsureRootPriv(cmd *cobra.Command, args []string) {
 			// The first argument is the context
 			sylog.Fatalf("command '%s %s' requires root privileges", args[0], cmd.Name())
 		} else {
-			sylog.Fatalf("command %s requires root privileges", cmd.Name())
+			sylog.Fatalf("command '%s' requires root privileges", cmd.CommandPath())
```
Comment (sashayakovtseva) [inline cmd/internal/cli/actions_linux.go]:
```suggestion
			sylog.Fatalf("command %q requires root privileges", cmd.CommandPath())
```

File: cmd/internal/cli/config_fakeroot_linux.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	Name:         "disable",
+	ShortHand:    "d",
+	Usage:        "disable a user fakeroot mapping entry preventing him to use the fakeroot feature",
+}
+
+// configFakerootCmd singularity config fakeroot
+var configFakerootCmd = &cobra.Command{
+	Args:                  cobra.ExactArgs(1),
+	DisableFlagsInUseLine: true,
+	PreRun:                func(cmd *cobra.Command, args []string) { EnsureRootPriv(cmd, nil) },
```
Comment (sashayakovtseva) [inline cmd/internal/cli/config_fakeroot_linux.go]:
```suggestion
	PreRun:                EnsureRootPriv,
```

File: cmd/internal/cli/config_fakeroot_linux.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+		username := args[0]
+
+		subUIDConfig, err := fakeroot.GetConfig(fakeroot.SubUIDFile, true, nil)
```
Comment (sashayakovtseva) [inline cmd/internal/cli/config_fakeroot_linux.go]: Maybe add `defer  subUIDConfig.Close()`? That will require to change fatals to smth different. 
Not sure what is better here.

File: cmd/internal/cli/config_fakeroot_linux.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+// -e|--enable
+var fakerootConfigEnable bool
+var fakerootConfigEnableFlag = cmdline.Flag{
+	ID:           "fakerootConfigEnableFlag",
+	Value:        &fakerootConfigEnable,
+	DefaultValue: false,
+	Name:         "enable",
+	ShortHand:    "e",
+	Usage:        "enable a user fakeroot mapping entry allowing him to use the fakeroot feature",
```
Comment (sashayakovtseva) [inline cmd/internal/cli/config_fakeroot_linux.go]: should we mention that a user must be added before enabling?

File: cmd/internal/cli/config_fakeroot_linux.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+// -d|--disable
+var fakerootConfigDisable bool
+var fakerootConfigDisableFlag = cmdline.Flag{
+	ID:           "fakerootConfigDisableFlag",
+	Value:        &fakerootConfigDisable,
+	DefaultValue: false,
+	Name:         "disable",
+	ShortHand:    "d",
+	Usage:        "disable a user fakeroot mapping entry preventing him to use the fakeroot feature",
```
Comment (sashayakovtseva) [inline cmd/internal/cli/config_fakeroot_linux.go]: should we mention a user should be added already?

File: docs/content.go [hunk 6]
```diff
@@ -966,4 +966,33 @@ Enterprise Performance Computing (EPC)`
   mount.`
 	OciUmountExample string = `
   $ singularity oci umount /var/lib/singularity/bundles/example`
+
+	ConfigUse   string = `config`
+	ConfigShort string = `Manage various singularity configuration (root user only)`
+	ConfigLong  string = `
+  The config command allow root user to manage various configuration like fakeroot
```
Comment (sashayakovtseva) [inline docs/content.go]:
```suggestion
  The config command allows root user to manage various configuration like fakeroot
```

File: internal/app/singularity/config_fakeroot_linux.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+	"github.com/sylabs/singularity/internal/pkg/fakeroot"
+)
+
+// FakerootConfigOp defines a type for a fakeroot
+// configuration operation.
+type FakerootConfigOp uint8
+
+const (
+	// FakerootAddUser is the operation to add a user fakeroot mapping
```
Comment (sashayakovtseva) [inline internal/app/singularity/config_fakeroot_linux.go]:
```suggestion
	// FakerootAddUser is the operation to add a user fakeroot mapping.
```

File: internal/app/singularity/config_fakeroot_linux.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	FakerootAddUser FakerootConfigOp = iota
+	// FakerootRemoveUser is the operation to remove a user fakeroot mapping
```
Comment (sashayakovtseva) [inline internal/app/singularity/config_fakeroot_linux.go]:
```suggestion
	// FakerootRemoveUser is the operation to remove a user fakeroot mapping.
```

File: internal/app/singularity/config_fakeroot_linux.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	FakerootRemoveUser
+	// FakerootEnableUser is the operation to enable a user fakeroot mapping
```
Comment (sashayakovtseva) [inline internal/app/singularity/config_fakeroot_linux.go]:
```suggestion
	// FakerootEnableUser is the operation to enable a user fakeroot mapping.
```

File: internal/app/singularity/config_fakeroot_linux.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	FakerootEnableUser
+	// FakerootDisableUser is the operation to disable a user fakeroot mapping
```
Comment (sashayakovtseva) [inline internal/app/singularity/config_fakeroot_linux.go]:
```suggestion
	// FakerootDisableUser is the operation to disable a user fakeroot mapping.
```

File: internal/app/singularity/config_fakeroot_linux.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	FakerootDisableUser
+)
+
+// FakerootConfig allows to add/remove/enable/disable a user fakeroot
+// mapping entry in /etc/subuid and /etc/subgid files.
+func FakerootConfig(username string, op FakerootConfigOp) error {
+	subUIDConfig, err := fakeroot.GetConfig(fakeroot.SubUIDFile, true, nil)
```
Comment (sashayakovtseva) [inline internal/app/singularity/config_fakeroot_linux.go:32]: Maybe closing a config should be in a defer call?

**Review by sashayakovtseva — CHANGES_REQUESTED** [review]

File: cmd/internal/cli/config_fakeroot_linux.go (hunk 3 above)
Comment (cclerget) [reply cmd/internal/cli/config_fakeroot_linux.go]: `Close` behind the scene flushes any pending update to the underlying `/etc/subXid` file (like after a `AddUser` call), so if we defer it and the call to `subGIDConfig.AddUser` fail for some reasons we will update `/etc/subuid` and maybe `/etc/subgid` with a wrong entry. No defer here, this is deliberate

File: internal/app/singularity/config_fakeroot_linux.go (hunk 11 above)
Comment (cclerget) [reply internal/app/singularity/config_fakeroot_linux.go:32]: same as above

**Review by sashayakovtseva — APPROVED** [review]

**Comment by DrDaveD** [comment]
This did not make it into the 3.4 release so please change the milestone.

**Pull request merged**